design/two-stage-backup-api-proposal.md

```diff
@@ -0,0 +1,149 @@
++----------+---------+
+           |
+           |
+           | < 2 >
+           |
+           v
++----------+---------+
+|                    |
+|    Post-Hook       |
+|                    |
++--------------------+
+
+
+
+
+```
+
+In the above flow, `CreateSnapshot` method creates snapshot and starts upload too. Hence, split this into two APIs: `CreateSnapshot` and `UploadSnapshot`.
+
+Proposed design: `Pre-hook` -> `CreateSnapshot` -> `Post-hook` -> `UploadSnapshot` (Background)
+
+```acsii
+
++--------------------+
+|                    |
+|     Pre-Hook       |
+|                    |
++---------+----------+
+          |
+          |
+          | < 1 >
+          |
+          v
++---------+----------+
+|                    |
+|   CreateSnapshot   |
+|                    |
++----------+---------+
+           |
+           |
+           | < 2 >
+           |
+           v
++----------+---------+
+|                    |
+|    Post-Hook       |
+|                    |
++----------+---------+
+           |
+           | < 3 >
+           |
+           v
+ +---------+---------+
+ |                   |
+ |  UploadSnapshot   |
+ |   (Background)    |
+ +-------------------+
+
+
+```
+
+## Detailed Design
+
+Velero has a pretty neat feature to maintain the application consistent backups/snapshot called `Pre-hooks` and `Post-Hooks`. However, there's a small change in Velero which needs to be addressed in order to achieve minimum downtime for an application or more consistent application backups. Existing backup flow for a typical stateful application looks like: `Pod` -> `Pre-Hook` --> `PVC` -> `PV` -> `Post-Hook`. This is a mechanism which is followed by Velero block store plugins. In most of the plugins, `CreateSnapshot` method of `VolumeSnapshotter` interface implements logic to upload the snapshot and this is a blocking call for `Post-Hook`. This will introduce more time and will keep application frozen for time equals to snapshot upload time.
+
+
+To solve the above problem, we can split existing backup flow and introduce one more method called `UploadSnapshot` to `VolumeSnapshotter` interface, which will execute once `Post-Hook` is executed and will keep running in background. This will reduce the time between `Pre-Hook` and `Post-Hook` significantly as well as it will not impact application consistency or downtime.
+
+### Implementation
+
+This change should be implemented in both Velero as well as Velero plugins. However, Velero plugins can continue to rely on current implementation. Only way for Velero plugins to get benifitted from this approach is to implement `UploadSnapshot` method.
+
+Volumesnapshotter is designed for snapshot, this can be either cloud snapshot or local snapshot. As of now, most plugin use volumesnapshotlocations to decide the snapshot type.
+We can add new config parameter in velero command line to define snapshot type, through this, velero will decide whether to invoke `UploadSnapshot` or not.
```

Comment:
This parameter is to decide whether Velero should execute `uploadSnapshot` or not.
There are few storage providers like openEBS, portworx, who support local snapshot and cloud base snapshot. With `uploadSnapshot` API velero will enable uploading of the snapshot(or cloud base snapshot).

What if the user wants a local snapshot only?

I agree that this can be achieved by volumeSnapshotLocation also, where the plugin will decide if it is configured to upload snapshot or not. Since velero is providing `uploadSnapshot` API,
it will be better to have this control in Velero itself. This parameter will go to `BackupSpec`.

> And, for thoroughness, is there a possibility for a need of the reverse, needing to call UploadSnapshot but not CreateSnapshot?

For this, the snapshot should exist. But it's up to plugin since `uploadSnapshot` consumes `snapshotId` to upload snapshot.